tests/unit/test_salesforce.py

```diff
@@ -0,0 +1,101 @@
+import pandas as pd
+import pytest
+
+from viadot.sources import Salesforce
+
+TABLE_TO_DOWNLOAD = "Account"
+TABLE_TO_UPSERT = "Contact"
+TEST_LAST_NAME = "prefect-viadot-test"
+ID_TO_UPSERT = "0035E00001YGWK3QAP"
+
+
+@pytest.fixture(scope="session")
+def salesforce():
+    s = Salesforce(config_key="sales_force_dev")
+    yield s
+
+
+@pytest.fixture(scope="session")
+def test_df_data(salesforce):
+    data = {
+        "Id": [ID_TO_UPSERT],
+        "LastName": [TEST_LAST_NAME],
+    }
+    df = pd.DataFrame(data=data)
+
+    yield df
+
+    data_restored = {
+        "Id": [ID_TO_UPSERT],
+        "LastName": ["LastName"],
+    }
+    df_restored = pd.DataFrame(data=data_restored)
+    salesforce.upsert(df=df_restored, table=TABLE_TO_UPSERT)
+
+
+@pytest.fixture(scope="session")
+def test_df_external(salesforce):
+    data = {
+        "LastName": [TEST_LAST_NAME],
+        "SAPContactId__c": ["111"],
+    }
+    df = pd.DataFrame(data=data)
+    yield df
+
+    data_restored = {
+        "Id": [ID_TO_UPSERT],
+        "LastName": ["LastName"],
+    }
+    df_restored = pd.DataFrame(data=data_restored)
+    salesforce.upsert(df=df_restored, table=TABLE_TO_UPSERT)
+
+
+def test_upsert_empty(salesforce):
+    try:
+        df = pd.DataFrame()
+        salesforce.upsert(df=df, table=TABLE_TO_UPSERT)
+    except Exception as exception:
+        assert False, exception
+
+
+def test_upsert_external_id_correct(salesforce, test_df_external):
+    try:
+        salesforce.upsert(
+            df=test_df_external, table=TABLE_TO_UPSERT, external_id="SAPContactId__c"
+        )
+    except Exception as exception:
+        assert False, exception
```

Comment:
don't retest what another test has already tested

Reply:
This test checks the upsert based on external_ID. Other tests perform an upsert based on the row ID

Reply:
I corrected the structure of this test
[✅ 🎨Improved structure of upsert test](https://github.com/dyvenia/viadot/pull/629/commits/de73e4ca66fa277fb67d44c7cd5221f0c9eacc00)